dlib/matrix/fft_size.h

```diff
@@ -0,0 +1,215 @@
+               - L.size() > 0 and L.size() <= 5
+               - L contains strictly positive values
+
+            Once the object is constructed, it is immutable.
+        !*/
+    private:
+        using container_type    = std::array<long,5>;
+        using const_reference   = container_type::const_reference;
+        using iterator          = container_type::iterator;
+        using const_iterator    = container_type::const_iterator;
+        
+        size_t _size = 0;
+        container_type _dims;
+        
+        iterator begin()
+        {
+            return _dims.begin();
+        }
+        
+        const_iterator begin() const
+        {
+            return _dims.begin();
+        }
+        
+        iterator end()
+        {
+            return _dims.begin() + _size;
+        }
+        
+        const_iterator end() const
+        {
+            return _dims.begin() + _size;
+        }
+
+    public:
+        fft_size() = default;
+        /*!
+            ensures
+                - this is properly initialised
+                - size() == 0
+                - is_valid() == false
+        !*/
+        
+        fft_size(std::initializer_list<long> dims)
+        {
+            DLIB_ASSERT(dims.size() > 0, "the initialiser list must be non-empty");
+            DLIB_ASSERT(dims.size() <= _dims.size(), "the initialiser list must have size less than 6");
+            std::copy(dims.begin(), dims.end(), _dims.begin());
+            _size = dims.size();
+            DLIB_ASSERT(is_valid(), "the initialiser list must contain strictly positive values");
+        }
+        /*!
+            requires
+                - dims.size() > 0 and dims.size() <= 5
+                - dims contains strictly positive values
+            ensures
+                - *this is properly initialised
+                - size() == dims.size()
+                - is_valid() == true 
+            throws
+                - dlib::fatal_error if requirements aren't satisfied.
+        !*/
+        
+        size_t size() const
+        {
+            return _size;
+        }
+        /*!
+            ensures
+                - returns the number of elements in *this
+        !*/
+
+        const_reference operator[](size_t index) const
+        {
+            DLIB_ASSERT(index < _size, "index " << index << " out of range [0," << _size << ")");
+            return _dims[index];
+        }
+        /*!
+            requires
+                - size() > 0
+                - index < size()
+            ensures
+                - returns a const reference to the element at position index
+        !*/
+        
+        const_reference back() const
+        {
+            DLIB_ASSERT(_size > 0, "object is empty");
+            return _dims[_size-1];
+        }
+        /*!
+            requires
+                - size() > 0
+            ensures
+                - returns a const reference to (*this)[size()-1]
+        !*/
+        
+        long dimprod() const
```

Comment:
It would be more clear if `size()` and `dimprod()` were renamed to `num_dims()` and `num_elements()` or something like that.

Reply:
okidok